File: android/app/src/main/java/at/kaufi/kantine/domain/splitter/LangModel.kt
package at.kaufi.kantine.domain.splitter

import kotlin.math.ln
import kotlin.math.max

/**
 * The trigram-based language model for distinguishing German from English.
 *
 * Pre-computes trigram log probabilities from [seed] data in [init].
 * Exposes [scorePhrase] and [scoreLang] matching the JS `createLangModel`.
 *
 * NOTE: The JS `learnFromCourse`, `saveDelta`, and `loadDelta` are intentionally
 * NOT ported — learning is a client-side feature not needed in the Android app.
 */
class LangModel(seed: LangModelSeed) {

    private val trigramsDe: Map<String, Int>
    private val trigramsEn: Map<String, Int>
    private val funcDe: Set<String>
    private val funcEn: Set<String>
    private val totalDe: Int
    private val totalEn: Int

    init {
        trigramsDe = seed.trigramsDe
        trigramsEn = seed.trigramsEn
        funcDe = seed.funcDe.toSet()
        funcEn = seed.funcEn.toSet()

        var td = 0
        for (v in trigramsDe.values) td += v
        totalDe = td

        var te = 0
        for (v in trigramsEn.values) te += v
        totalEn = te
    }

    /**
     * Score [text] for German vs English trigram affinity.
     * Returns a pair of (deScore, enScore).
     */
    fun scorePhrase(text: String?): Pair<Double, Double> {
        if (text.isNullOrBlank()) return Pair(0.0, 0.0)

        val lowerText = text.lowercase()
        val alphaWords = Regex("[a-zäöüß]+").findAll(lowerText).map { it.value }.toList()

        val FUNC_WEIGHT = 2.0

        var deTriLog = 0.0
        var enTriLog = 0.0

        for (w in alphaWords) {
            for (i in 0..w.length - 3) {
                val tri = w.substring(i, i + 3)

                val countDe = trigramsDe[tri] ?: 0
                deTriLog += ln((countDe + 1).toDouble() / (totalDe + 2).toDouble())

                val countEn = trigramsEn[tri] ?: 0
                enTriLog += ln((countEn + 1).toDouble() / (totalEn + 2).toDouble())
            }
        }

        val minTri = minOf(deTriLog, enTriLog)
        deTriLog -= minTri
        enTriLog -= minTri

        var deScore = deTriLog
        var enScore = enTriLog

        for (w in alphaWords) {
            if (w in funcDe) deScore += FUNC_WEIGHT
            if (w in funcEn) enScore += FUNC_WEIGHT
        }

        val umlauts = Regex("[äöüß]").findAll(lowerText).toList()
        if (umlauts.isNotEmpty()) {
            deScore += 0.5 * umlauts.size
        }

        for (w in alphaWords) {
            if (Regex("(ung|suppe|chen|kartoffel|schnitzel)$").containsMatchIn(w)) deScore += 1.0
            if (Regex("(ing|ed)$").containsMatchIn(w)) enScore += 0.5
            if (Regex("^th").containsMatchIn(w)) enScore += 0.5
        }

        val deDigraphs = Regex("(sch|pf|tz|ck)").findAll(lowerText).toList()
        if (deDigraphs.isNotEmpty()) {
            deScore += 0.3 * deDigraphs.size
        }

        return Pair(deScore, enScore)
    }

    /**
     * Score language: positive = German, negative = English.
     */
    fun scoreLang(text: String?): Double {
        val scores = scorePhrase(text)
        return scores.first - scores.second
    }
}

File: src/lang/langModel.js
import { isLoanword } from './loanwords.js';

export function createLangModel(seed) {
    const FUNC_WEIGHT = 2.0;

    const trigramsDe = { ...(seed.trigramsDe || {}) };
    const trigramsEn = { ...(seed.trigramsEn || {}) };
    const funcDe = new Set(seed.funcDe || []);
    const funcEn = new Set(seed.funcEn || []);

    let totalDe = 0;
    for (const k in trigramsDe) {
        totalDe += trigramsDe[k];
    }

    let totalEn = 0;
    for (const k in trigramsEn) {
        totalEn += trigramsEn[k];
    }

    function scorePhrase(text) {
        if (!text) return { de: 0, en: 0 };

        let deScore = 0;
        let enScore = 0;

        const lowerText = text.toLowerCase();
        // Filter out loanword tokens: cross-lingual food terms (lasagne, gnocchi,
        // schnitzel, ...) appear in BOTH German and English descriptions, so they
        // should not bias the DE/EN evidence either way. Same applies to global
        // text rules (umlauts, digraphs) — "schnitzel" must not trigger the
        // "sch"+"tz" DE digitraph bonus.
        const alphaWords = (lowerText.match(/[a-zäöüß]+/g) || []).filter(w => !isLoanword(w));
        const filteredText = alphaWords.join(' ');

        let deTriLog = 0;
        let enTriLog = 0;

        for (const w of alphaWords) {
            for (let i = 0; i <= w.length - 3; i++) {
                const tri = w.substring(i, i + 3);

                const countDe = trigramsDe[tri] || 0;
                deTriLog += Math.log((countDe + 1) / (totalDe + 2));

                const countEn = trigramsEn[tri] || 0;
                enTriLog += Math.log((countEn + 1) / (totalEn + 2));
            }
        }

        const minTri = Math.min(deTriLog, enTriLog);
        deScore += (deTriLog - minTri);
        enScore += (enTriLog - minTri);

        for (const w of alphaWords) {
            if (funcDe.has(w)) deScore += FUNC_WEIGHT;
            if (funcEn.has(w)) enScore += FUNC_WEIGHT;
        }

        const umlauts = filteredText.match(/[äöüß]/g);
        if (umlauts) {
            deScore += 0.5 * umlauts.length;
        }

        for (const w of alphaWords) {
            if (/(ung|suppe|chen|kartoffel|schnitzel)$/.test(w)) deScore += 1.0;
            if (/(ing|ed)$/.test(w)) enScore += 0.5;
            if (/^th/.test(w)) enScore += 0.5;
        }

        const deDigraphs = filteredText.match(/(sch|pf|tz|ck)/g);
        if (deDigraphs) {
            deScore += 0.3 * deDigraphs.length;
        }

        return { de: deScore, en: enScore };
    }

    function scoreLang(text) {
        const scores = scorePhrase(text);
        return scores.de - scores.en;
    }

    function scoreCharAffinities(text) {
        if (!text) return [];

        const lowerText = text.toLowerCase();
        const len = lowerText.length;
        const rawScores = new Array(len).fill(0);
        const counts = new Array(len).fill(0);

        for (let i = 0; i <= len - 3; i++) {
            const tri = lowerText.substring(i, i + 3);
            
            if (!/^[a-zäöüß]{3}$/.test(tri)) continue;

            const countDe = trigramsDe[tri] || 0;
            const countEn = trigramsEn[tri] || 0;
            const logDe = Math.log((countDe + 1) / (totalDe + 2));
            const logEn = Math.log((countEn + 1) / (totalEn + 2));
            const signedDiff = logDe - logEn;

            for (let j = 0; j < 3; j++) {
                rawScores[i + j] += signedDiff;
                counts[i + j]++;
            }
        }

        const averaged = rawScores.map((sum, i) => counts[i] > 0 ? sum / counts[i] : 0);

        const maxAbs = Math.max(...averaged.map(Math.abs), 1e-9);
        const normalized = averaged.map(v => v / maxAbs);

        const result = [];
        for (let i = 0; i < len; i++) {
            result.push({
                char: text[i],
                affinity: normalized[i]
            });
        }

        return result;
    }

    function getModel() {
        return {
            version: seed.version,
            trigramsDe: trigramsDe,
            trigramsEn: trigramsEn,
            funcDe: Array.from(funcDe),
            funcEn: Array.from(funcEn)
        };
    }

    const modelObj = {
        scorePhrase,
        scoreLang,
        scoreCharAffinities,
        getModel
    };

    return modelObj;
}
